optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -180,7 +190,332 @@
      * @return a stream that matches every combination of A and B for which all the {@link BiJoiner joiners} are true
      */
     default <B> BiConstraintStream<A, B> join(Class<B> otherClass, BiJoiner<A, B>... joiners) {
-        return join(otherClass, AbstractBiJoiner.merge(joiners));
+        int joinerCount = joiners.length;
+        int indexOfFirstFilter = -1;
+        for (int i = 0; i < joinerCount; i++) {
+            BiJoiner<A, B> joiner = joiners[i];
+            boolean hasAFilter = indexOfFirstFilter >= 0;
+            if (joiner instanceof FilteringBiJoiner) {
+                if (!hasAFilter) { // From now on, we only allow filtering joiners.
+                    indexOfFirstFilter = i;
+                }
+            } else if (hasAFilter) {
+                throw new IllegalStateException("Indexing joiner (" + joiner + ") must not follow a filtering joiner ("
```

Comment:
This exception and the expected order of joiners should be probably described in a javadoc.

This fail-fast is not a super trivial check of inputs - it establishes a contract regarding in which order users are supposed to register joiners of different kinds. There should be a test for it.